### PR TITLE
[#589] PodDisruptionBudget is not part of the deployed resources list

### PR DESCRIPTION
**Pull request by gaohoward**

### CODE REVIEW

File: controllers/activemqartemis_controller.go [hunk 1]
```diff
@@ -121,7 +121,7 @@ type ActiveMQArtemisReconciler struct {
 //+kubebuilder:rbac:groups=monitoring.coreos.com,namespace=activemq-artemis-operator,resources=servicemonitors,verbs=get;create
 //+kubebuilder:rbac:groups=apps,namespace=activemq-artemis-operator,resources=deployments/finalizers,verbs=update
 //+kubebuilder:rbac:groups=rbac.authorization.k8s.io,namespace=activemq-artemis-operator,resources=roles;rolebindings,verbs=create;get;delete
-//+kubebuilder:rbac:groups=policy,namespace=activemq-artemis-operator,resources=poddisruptionbudgets,verbs=create;get;delete
+//+kubebuilder:rbac:groups=policy,namespace=activemq-artemis-operator,resources=poddisruptionbudgets,verbs=create;get;delete;list;watch
```
Comment (gtully) [inline controllers/activemqartemis_controller.go:124]: I see that we need the list permission, not sure that we need the watch? We won't watch for changes, but we do want to be able to update

File: controllers/activemqartemis_controller_test.go [hunk 2]
```diff
@@ -442,10 +442,26 @@ var _ = Describe("artemis controller", func() {
 			pdbObject := policyv1.PodDisruptionBudget{}
 			Eventually(func(g Gomega) {
 				g.Expect(k8sClient.Get(ctx, pdbKey, &pdbObject)).Should(Succeed())
+
+				resMap, err := getDeployedResources(createdCr, k8sClient)
```
Comment (gtully) [inline controllers/activemqartemis_controller_test.go:446]: I think this use of getDeployedResources is too low level. Ideally these tests just work on the kube api.
A test that verifies that we can update the PDB when the CR updates may be at a better level of abstraction. Without the fix, an update will always fail b/c we see it as a create in error. 
A test that creates a CR with PDB, then updates the CR with a change that is relevant to the PDB and verifies that the PDB resource version is updated (!= to the original) will cover this nicely.

File: deploy/cluster_role.yaml [hunk 3]
```diff
@@ -174,6 +174,8 @@ rules:
   - create
   - delete
   - get
+  - list
+  - watch
```
Comment (gtully) [inline deploy/cluster_role.yaml:178]: not seeing the need for watch!

**Review by gtully — CHANGES_REQUESTED** [review]

File: controllers/activemqartemis_controller.go (hunk 1 above)
Comment (gaohoward) [reply controllers/activemqartemis_controller.go:124]: Yes that's strange if I don't add the `watch` permission, the operator seems work but the log will show some warning message regarding this permission periodically.

File: controllers/activemqartemis_controller_test.go (hunk 2 above)
Comment (gaohoward) [reply controllers/activemqartemis_controller_test.go:446]: yeah make sense. will add it.

File: deploy/cluster_role.yaml (hunk 3 above)
Comment (gaohoward) [reply deploy/cluster_role.yaml:178]: Not sure why as replied above, I need more investigation.

Comment (gaohoward) [reply deploy/cluster_role.yaml:179]: yes I think we need the watch permission. If I remove the watch, the pdb created the first time are not quickly sync'ed in the cache so the controller take a long time to know that there is already a pdb created (so the controller will try a few times to create it and get 'already exists' error in each following reconcile loop).

File: controllers/activemqartemis_controller_test.go [hunk 4]
```diff
@@ -485,6 +486,57 @@ var _ = Describe("artemis controller", func() {
...
+
+			Eventually(func(g Gomega) {
+				g.Expect(k8sClient.Get(ctx, pdbKey, &pdbObject)).Should(Succeed())
+				g.Expect(*pdbObject.Spec.MinAvailable).To(BeEquivalentTo(intstr.FromInt(2)))
+				g.Expect(prevResourceVersion).NotTo(Equal(pdbObject.ResourceVersion))
+			}, existingClusterTimeout, existingClusterInterval).Should(Succeed())
+
+			CleanResource(brokerCr, brokerCr.Name, defaultNamespace)
+		})
 	})
```
Comment (gtully) [inline controllers/activemqartemis_controller_test.go:540]: nice test!

File: deploy/cluster_role.yaml [hunk 5]
```diff
@@ -174,6 +174,9 @@ rules:
   - create
   - delete
   - get
+  - list
+  - update
+  - watch
```
Comment (gtully) [inline deploy/cluster_role.yaml:179]: still don't see the need for the watch permission, is there a way to remove?

**Review by gtully — APPROVED** [review]

File: deploy/cluster_role.yaml (hunk 5 above)
Comment (gaohoward) [reply deploy/cluster_role.yaml:179]: We are not directly use it. It is used for cache update I think.

File: controllers/activemqartemis_controller.go [hunk 6]
```diff
@@ -121,7 +121,7 @@ type ActiveMQArtemisReconciler struct {
 //+kubebuilder:rbac:groups=monitoring.coreos.com,namespace=activemq-artemis-operator,resources=servicemonitors,verbs=get;create
 //+kubebuilder:rbac:groups=apps,namespace=activemq-artemis-operator,resources=deployments/finalizers,verbs=update
 //+kubebuilder:rbac:groups=rbac.authorization.k8s.io,namespace=activemq-artemis-operator,resources=roles;rolebindings,verbs=create;get;delete
-//+kubebuilder:rbac:groups=policy,namespace=activemq-artemis-operator,resources=poddisruptionbudgets,verbs=create;get;delete
+//+kubebuilder:rbac:groups=policy,namespace=activemq-artemis-operator,resources=poddisruptionbudgets,verbs=create;get;delete;list;update;watch
```
Comment (gtully) [inline controllers/activemqartemis_controller.go:124]: this watch permission. I don't think that is necessary. Is it? we should not watch it b/c we already watch the CR.

Comment (gaohoward) [reply controllers/activemqartemis_controller.go:124]: this watch permission is needed, but not directly. If the permission is absent, the operator got issues with client cache updates. (i.e. when you create the resource the cache won't get updated quickly). (same thing with other resources like services which we don't directly watch them).

**Pull request merged**